orttraining/orttraining/eager/opgen/opgen/generator.py

```diff
@@ -294,244 +476,192 @@ def _write_function_body(self, writer: opgenwriter.SourceWriter, mapped_func: Ma
-                writer.writeline(f"std::move({op_input}),")
-            writer.pop_indent()
-            writer.write(f"}}, {onnx_op.outputs}, {attrs_arg}")
-            if onnx_op.domain:
-                writer.write(f", {onnx_op.domain}")
-            writer.writeline(");")
-            writer.writeline("CHECK_STATUS(status);")
-            writer.writeline()
+        if return_info:
+            self._write_function_body_return_info_outputs(
+                writer, in_place_params, cpp_func, return_info, onnx_op, onnx_op_index
+            )
+            self._write_function_body_set_out(
+                writer,
+                in_place_params,
+                set_out_tensor,
+                onnx_op_index,
+                ctx,
+                need_type_promotion,
+                cast_op_found,
+                onnx_op,
+                return_info,
+                cpp_func,
+            )
 
-            # We'll potentially return back to Torch from this op
-            return_outputs = onnx_op.outputs
+        # We'll potentially return back to Torch from this op
+        return_outputs = self._write_function_body_onnx_op_invocation(
+            writer, onnx_op, onnx_op_index, cpp_func, attrs_arg_ptr
+        )
 
-        # TODO: Pick the right "out" Torch parameter; do not assume the first one
-        # TODO: Handle multiple results
-        # TODO: Assert return type
+        return in_place_params, return_outputs
 
-        if cpp_func.return_type.desugar().identifier_tokens[0].value == "void":
-            pass
-        elif len(in_place_params) == 0:
-            # tensor options
-            if set_out_tensor:
-                if need_type_promotion and not cast_op_found:
-                    writer.writeline("if (*promoted_type != out.scalar_type()) {")
-                    writer.push_indent()
-                    writer.writeline(
-                        f"CastToType_out(invoker, {onnx_op.outputs}[0], ort_input_out, out.scalar_type());"
-                    )
-                    writer.pop_indent()
-                    writer.writeline("}")
-
-                writer.writeline(f"return {last_param.identifier.value};")
-                return
-
-            # TODO: revisit the hardcoded use of TensorList.
-            writer.write(f"at::TensorOptions tensor_options = {first_param.identifier.value}")
-            if first_param.parameter_type.desugar().identifier_tokens[0].value == "TensorList":
-                writer.write("[0]")
-            writer.write(".options()")
-            if need_type_promotion:
-                writer.write(".dtype(*promoted_type)")
+    # TODO: Pick the right "out" Torch parameter; do not assume the first one
+    # TODO: Handle multiple results
+    # TODO: Assert return type
+    # TODO: warn if we have not consumed all torch parameters (either as an ORT input or ORT attribute).
+    def _write_function_body(self, writer: opgenwriter.SourceWriter, mapped_func: MappedOpFunction):
```

Comment:
Reading through these changes - as we are refactoring this file into smaller / logically scoped functions, I was almost expecting a top level `_write_function(...)` function to start from... To logically structure the function that generates the function.

Reply:
This would have to be inside the for loop in `run`, but only a little bit of the code there would actually move into a `_write_function` method, as the rest is dealing with namespaces and/or tracking the list of functions.  I could be persuaded to extract the signature + body calls to a separate one if you think it's clearer that way.